app/product_notices.go

```diff
@@ -0,0 +1,260 @@
+
+	// check if client version is in notice range
+	clientVersions := cnd.DesktopVersion
+	if client == model.NoticeClientType_MobileAndroid || client == model.NoticeClientType_MobileIos {
+		clientVersions = cnd.MobileVersion
+	}
+
+	clientVersionParsed, err := semver.NewVersion(clientVersion)
+	if err != nil {
+		return false, errors.Wrapf(err, "Cannot parse version range %s", clientVersion)
+	}
+
+	for _, v := range clientVersions {
+		c, err := semver.NewConstraint(v)
+		if err != nil {
+			return false, errors.Wrapf(err, "Cannot parse version range %s", v)
+		}
+		if !c.Check(clientVersionParsed) {
+			return false, nil
+		}
+	}
+
+	// check if notice date range matches current
+	if cnd.DisplayDate != nil {
+		now := time.Now().UTC()
+		c, err := date_constraints.NewConstraint(*cnd.DisplayDate)
+		if err != nil {
+			return false, errors.Wrapf(err, "Cannot parse date range %s", *cnd.DisplayDate)
+		}
+		if !c.Check(&now) {
+			return false, nil
+		}
+	}
+
+	// check if current server version is notice range
+	serverVersion, _ := semver.NewVersion(model.CurrentVersion)
+	for _, v := range cnd.ServerVersion {
+		c, err := semver.NewConstraint(v)
+		if err != nil {
+			return false, errors.Wrapf(err, "Cannot parse version range %s", v)
+		}
+		if !c.Check(serverVersion) {
+			return false, nil
+		}
+	}
+
+	// check if sku matches our license
+	if cnd.Sku != nil {
+		if !cnd.Sku.Matches(sku) {
+			return false, nil
+		}
+	}
+
+	// check the target audience
+	if cnd.Audience != nil {
+		if !cnd.Audience.Matches(isSystemAdmin, isTeamAdmin) {
+			return false, nil
+		}
+
+	}
+
+	// check user count condition against previously calculated total user count
+	if cnd.NumberOfUsers != nil && userCount > 0 {
+		if userCount < *cnd.NumberOfUsers {
+			return false, nil
+		}
+	}
+
+	// check post count condition against previously calculated total post count
+	if cnd.NumberOfPosts != nil && postCount > 0 {
+		if postCount < *cnd.NumberOfPosts {
+			return false, nil
+		}
+	}
+
+	// check if our server config matches the notice
+	for k, v := range cnd.ServerConfig {
+		if !validateConfigEntry(a, k, v) {
+			return false, nil
+		}
+	}
+
+	// check the type of installation
+	if cnd.InstanceType != nil {
+		if !cnd.InstanceType.Matches(isCloud) {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func validateConfigEntry(a *App, path string, expectedValue interface{}) bool {
+	value, found := config.GetConfigValueByPath(a.Config(), strings.Split(path, "."))
+	if !found {
+		return false
+	}
+	vt := reflect.ValueOf(value)
+	if vt.IsNil() {
+		return expectedValue == nil
+	}
+	if vt.Kind() == reflect.Ptr {
+		vt = vt.Elem()
+	}
+	val := vt.Interface()
+	return val == expectedValue
+}
+
+func (a *App) GetProductNotices(lastViewed int64, userId, teamId string, client model.NoticeClientType, clientVersion string, locale string) (model.NoticeMessages, *model.AppError) {
+	isSystemAdmin := a.SessionHasPermissionTo(*a.Session(), model.PERMISSION_MANAGE_SYSTEM)
```

Comment:
I don't think this is strictly wrong, but I don't know if this is strictly right either. This is saying that any user that has permissions to manage the system is a system admin, and any user with permissions to manage a team is a team admin. And that is kinda right. But maybe is bit more "right" to say that any user with the "system_admin" role is a system admin, and any user with the "team_admin" role is a team admin. From the notifications point of view, maybe makes more sense one or the other. probably is a question for PMs. The current approach is perfectly fine, but I think the other approach too.

Reply:
No idea which is the righter one :) Pinging @esethna

Reply:
For the purposes of notices we want to look at the user role. ie any user with the "system_admin" role is a system admin, and any user with the "team_admin" role is a team admin @reflog